<h1> Task 1 - Market Basket Analysis and Product Recommendation  </h1>

<h6> GOAL: The goal of this task is to use Spark MLlib to build a model to generate association rules to quickly run the market basket analysis to uncover associations between different items, then further to provide recommendations for purchase on a distributed platform.   </h6>

In [0]:
# Importing the library
from pyspark.ml.fpm import FPGrowth

In [0]:
# 1.3 - Creating the training set
# Reading the dataframe
train_df = spark.read.format("csv").option("header", "false").load("dbfs:/FileStore/shared_uploads/clvrashmika@gmail.com/Lab5_Part1_TrainData.csv")

In [0]:
# Displaying the dataframe
display(train_df)

_c0 _c1 _c2 _c3 _c4 bread null null null null peanut butter apple null null null peanut butter bread null null null peanut butter bread apple null null peanut butter bread milk null null peanut butter bread milk chocolate null bread milk orange null null apple chocolate milk null null peanut butter milk chocolate apple null cheese bread milk potatoes null cheese pasta ketchup null null milk cheese pasta ketchup null pasta ketchup cheese potatoes milk bread milk chocolate pasta null milk pasta potatoes ketchup bread apple chocolate pasta null null milk bread null null null apple milk null null null milk chocolate null null null milk null null null null

In [0]:
# Modifying the dataframe to perform analysis
count = 0
mylist = []
for row in train_df.collect():
    temp = (count,list(filter(None, row)))
    mylist.append(temp)
    count = count + 1

In [0]:
# Printing the list to show the new format
mylist

Out[122]: [(0, ['bread']),
 (1, ['peanut butter', ' apple']),
 (2, [' peanut butter', ' bread']),
 (3, ['peanut butter', ' bread', ' apple']),
 (4, ['peanut butter', ' bread', ' milk']),
 (5, ['peanut butter', ' bread', ' milk', ' chocolate']),
 (6, ['bread', ' milk', ' orange']),
 (7, ['apple', ' chocolate', ' milk']),
 (8, ['peanut butter', ' milk', ' chocolate', ' apple']),
 (9, ['cheese', ' bread', ' milk', ' potatoes ']),
 (10, ['cheese', ' pasta', ' ketchup']),
 (11, ['milk', ' cheese', ' pasta', ' ketchup']),
 (12, ['pasta', ' ketchup', ' cheese', ' potatoes', ' milk']),
 (13, ['bread', ' milk', ' chocolate', ' pasta']),
 (14, [' milk', ' pasta', ' potatoes', ' ketchup', ' bread']),
 (15, ['apple', ' chocolate', ' pasta']),
 (16, ['milk', 'bread']),
 (17, ['apple', 'milk']),
 (18, ['milk', 'chocolate']),
 (19, ['milk'])]

In [0]:
# Creating a new dataframe from the above list
train_df1 = spark.createDataFrame(mylist, ["id", "items"])

In [0]:
# 1.4 - FP Growth Model
fpGrowth = FPGrowth(itemsCol="items", minSupport=0.05, minConfidence=0.07)
model = fpGrowth.fit(train_df1)

In [0]:
#1.5 -  Display frequent itemsets.
model.freqItemsets.show()

+--------------------+----+
|               items|freq|
+--------------------+----+
|            [ apple]|   3|
|[ apple, peanut b...|   3|
|[ apple, peanut b...|   1|
|[ apple, peanut b...|   1|
|[ apple, peanut b...|   1|
|[ apple, peanut b...|   1|
|    [ apple,  bread]|   1|
|[ apple,  chocolate]|   1|
|[ apple,  chocola...|   1|
|     [ apple,  milk]|   1|
|             [pasta]|   1|
|   [pasta,  ketchup]|   1|
|[pasta,  ketchup,...|   1|
|  [pasta,  potatoes]|   1|
|[pasta,  potatoes...|   1|
|[pasta,  potatoes...|   1|
|[pasta,  potatoes...|   1|
|    [pasta,  cheese]|   1|
|[pasta,  cheese, ...|   1|
|[pasta,  cheese, ...|   1|
+--------------------+----+
only showing top 20 rows



In [0]:
#1.6  Display generated association rules.
display(model.associationRules)

antecedent consequent confidence lift support List( potatoes , milk) List(cheese) 1.0 10.0 0.05 List( potatoes , milk) List( bread) 1.0 3.3333333333333335 0.05 List( potatoes ) List(cheese) 1.0 10.0 0.05 List( potatoes ) List( bread) 1.0 3.3333333333333335 0.05 List( potatoes ) List( milk) 1.0 2.2222222222222223 0.05 List( potatoes) List(pasta) 0.5 10.0 0.05 List( potatoes) List( pasta) 0.5 2.0 0.05 List( potatoes) List( ketchup) 1.0 5.0 0.1 List( potatoes) List( bread) 0.5 1.6666666666666667 0.05 List( potatoes) List( milk) 1.0 2.2222222222222223 0.1 List( potatoes) List( cheese) 0.5 5.0 0.05 List(milk) List(apple) 0.2 1.3333333333333335 0.05 List(milk) List( pasta) 0.2 0.8 0.05 List(milk) List( cheese) 0.2 2.0 0.05 List(milk) List(chocolate) 0.2 4.0 0.05 List(milk) List(bread) 0.2 1.0 0.05 List(milk) List( ketchup) 0.2 1.0 0.05 List(bread, pasta) List( chocolate) 1.0 4.0 0.05 List(bread, pasta) List( milk) 1.0 2.2222222222222223 0.05 List(peanut butter, chocolate) List( apple) 0.5 3.3333333333333335 0.05 List(peanut butter, chocolate) List( bread) 0.5 1.6666666666666667 0.05 List(peanut butter, chocolate) List( milk) 1.0 2.2222222222222223 0.1 List( chocolate, milk) List( apple) 0.25 1.6666666666666667 0.05 List( chocolate, milk) List(apple) 0.25 1.6666666666666667 0.05 List( chocolate, milk) List( bread) 0.25 0.8333333333333334 0.05 List( chocolate, milk) List( pasta) 0.25 1.0 0.05 List( chocolate, milk) List(peanut butter) 0.5 2.0 0.1 List( chocolate, milk) List(bread) 0.25 1.25 0.05 List(peanut butter, chocolate, milk) List( apple) 0.5 3.3333333333333335 0.05 List(peanut butter, chocolate, milk) List( bread) 0.5 1.6666666666666667 0.05 List(pasta, cheese, ketchup) List( milk) 1.0 2.2222222222222223 0.05 List(pasta, cheese, ketchup) List( potatoes) 1.0 10.0 0.05 List( ketchup) List(pasta) 0.25 5.0 0.05 List( ketchup) List( potatoes) 0.5 5.0 0.1 List( ketchup) List(cheese) 0.25 2.5 0.05 List( ketchup) List( cheese) 0.5 5.0 0.1 List( ketchup) List(milk) 0.25 1.0 0.05 List( ketchup) List( pasta) 0.75 3.0 0.15 List( ketchup) List( bread) 0.25 0.8333333333333334 0.05 List( ketchup) List( milk) 0.5 1.1111111111111112 0.1 List( cheese) List(pasta) 0.5 10.0 0.05 List( cheese) List(milk) 0.5 2.0 0.05 List( cheese) List( pasta) 0.5 2.0 0.05 List( cheese) List( ketchup) 1.0 5.0 0.1 List( cheese) List( potatoes) 0.5 5.0 0.05 List( cheese) List( milk) 0.5 1.1111111111111112 0.05 List(pasta, ketchup) List( milk) 1.0 2.2222222222222223 0.05 List(pasta, ketchup) List( potatoes) 1.0 10.0 0.05 List(pasta, ketchup) List( cheese) 1.0 10.0 0.05 List(pasta, cheese, potatoes, milk) List( ketchup) 1.0 5.0 0.05 List(pasta, cheese, ketchup, milk) List( potatoes) 1.0 10.0 0.05 List(cheese) List( pasta) 0.5 2.0 0.05 List(cheese) List( ketchup) 0.5 2.5 0.05 List(cheese) List( bread) 0.5 1.6666666666666667 0.05 List(cheese) List( milk) 0.5 1.1111111111111112 0.05 List(cheese) List( potatoes ) 0.5 10.0 0.05 List(pasta, cheese) List( ketchup) 1.0 5.0 0.05 List(pasta, cheese) List( potatoes) 1.0 10.0 0.05 List(pasta, cheese) List( milk) 1.0 2.2222222222222223 0.05 List( cheese, ketchup) List(pasta) 0.5 10.0 0.05 List( cheese, ketchup) List(milk) 0.5 2.0 0.05 List( cheese, ketchup) List( pasta) 0.5 2.0 0.05 List( cheese, ketchup) List( milk) 0.5 1.1111111111111112 0.05 List( cheese, ketchup) List( potatoes) 0.5 5.0 0.05 List(apple) List(milk) 0.3333333333333333 1.3333333333333333 0.05 List(apple) List( pasta) 0.3333333333333333 1.3333333333333333 0.05 List(apple) List( chocolate) 0.6666666666666666 2.6666666666666665 0.1 List(apple) List( milk) 0.3333333333333333 0.7407407407407407 0.05 List( apple, milk) List(peanut butter) 1.0 4.0 0.05 List( apple, milk) List( chocolate) 1.0 4.0 0.05 List( chocolate, bread) List( milk) 1.0 2.2222222222222223 0.05 List( chocolate, bread) List(peanut butter) 1.0 4.0 0.05 List( cheese, ketchup, pasta) List(milk) 1.0 4.0 0.05 List( peanut butter) List( bread) 1.0 3.3333333333333335 0.05 List(bread, milk) List( pasta) 0.5 2.0

In [0]:
# 1.7 Creating a test set
test_df = spark.createDataFrame([
    (0, ['bread']),
    (1, ['potatoes','milk']),
    (2, ['chocolate']),
    (3, ['pasta', 'cheese']),
    (4, ['apple','milk']),
    (5, ['milk']),
    (6, ['chocolate', 'bread', 'milk']),
    (7, ['bread', 'milk'])
], ["id", "items"])

In [0]:
# 1.8 Making predictions
# transform examines the input items against all the association rules and summarize the consequents as prediction
display(model.transform(test_df))

id items prediction 0 List(bread) List(milk, pasta, chocolate, milk, orange) 1 List(potatoes, milk) List(apple, pasta, cheese, chocolate, bread, ketchup) 2 List(chocolate) List(milk) 3 List(pasta, cheese) List( pasta, ketchup, bread, milk, potatoes , potatoes, cheese) 4 List(apple, milk) List( pasta, cheese, chocolate, bread, ketchup, chocolate, milk) 5 List(milk) List(apple, pasta, cheese, chocolate, bread, ketchup) 6 List(chocolate, bread, milk) List(apple, pasta, cheese, ketchup, chocolate, milk, orange) 7 List(bread, milk) List(apple, pasta, cheese, chocolate, ketchup, chocolate, milk, orange)

#####  1.10 Task 1 - Additional

In [0]:
# Reading and displaying the dataframe
extra_df = spark.read.format("csv").option("header", "false").load("dbfs:/FileStore/shared_uploads/clvrashmika@gmail.com/groceries.csv")
display(extra_df)

_c0 _c1 _c2 _c3 citrus fruit semi-finished bread margarine ready soups tropical fruit yogurt coffee null whole milk null null null pip fruit yogurt cream cheese meat spreads other vegetables whole milk condensed milk long life bakery product whole milk butter yogurt rice rolls/buns null null null other vegetables UHT-milk rolls/buns bottled beer pot plants null null null whole milk cereals null null tropical fruit other vegetables white bread bottled water citrus fruit tropical fruit whole milk butter beef null null null frankfurter rolls/buns soda null chicken tropical fruit null null butter sugar fruit/vegetable juice newspapers fruit/vegetable juice null null null packaged fruit/vegetables null null null chocolate null null null specialty bar null null null other vegetables null null null butter milk pastry null null whole milk null null null tropical fruit cream cheese processed cheese detergent tropical fruit root vegetables other vegetables frozen dessert bottled water canned beer null null yogurt null null null sausage rolls/buns soda chocolate other vegetables null null null brown bread soda fruit/vegetable juice canned beer yogurt beverages bottled water specialty bar hamburger meat other vegetables rolls/buns spices root vegetables other vegetables whole milk beverages pork berries other vegetables whole milk beef grapes detergent null pastry soda null null fruit/vegetable juice null null null canned beer null null null root vegetables other vegetables whole milk dessert citrus fruit zwieback newspapers null sausage rolls/buns soda canned beer tropical fruit root vegetables whole milk yogurt berries yogurt null null canned beer null null null butter milk yogurt cream cheese spread cheese coffee null null null pastry bottled water null null rolls/buns null null null misc. beverages null null null root vegetables other vegetables butter curd sausage rolls/buns cat food newspapers canned beer null null null ham grapes other vegetables whole milk turkey tropical fruit other vegetables curd whole milk yogurt processed cheese pickled vegetables whole milk curd yogurt pastry packaged fruit/vegetables brown bread canned beer null rolls/buns oil bottled water chewing gum ham beef whipped/sour cream ice cream rolls/buns pastry sugar null other vegetables whole milk frozen vegetables canned fish sausage pastry null null sausage beef whole milk null frankfurter tropical fruit rolls/buns brown bread rolls/buns pastry soda null whole milk null null null curd cheese coffee null null red/blush wine newspapers null null sausage whole milk curd null tropical fruit pip fruit berries whole milk red/blush wine null null null whole milk butter margarine specialty fat frankfurter citrus fruit whole milk domestic eggs whole milk meat spreads soda null frozen potato products null null null other vegetables whole milk rolls/buns sugar citrus fruit whole milk curd butter milk flour salt bottled water fruit/vegetable juice sugar bottled water soda bottled beer frozen meals coffee null null chocolate null null null root vegetables whole milk frozen vegetables domestic eggs root vegetables onions hard cheese frozen vegetables herbs condensed milk frozen vegetables salt bottled water null null null sausage root vegetables onions yogurt coffee newspapers null null pork other vegetables whipped/sour cream null berries yogurt null null soda null null null berries null null null fruit/vegetable juice salty snack candles null citrus fruit butter milk yogurt cream cheese pork hamburger meat citrus fruit berries detergent null null null grapes photo/film null null sausage sliced cheese rolls/buns brown bread chicken hamburger meat citrus fruit tropical fruit whole milk yogurt brown bread null citrus fruit tropical fruit null null soda misc. beverages null null sausage pork grapes whole milk frankfurter rolls/buns bottled water null sausage whole milk yogurt coffee soda null null null rolls/buns fruit/vegetable juice null null beef citrus fruit berries ro

In [0]:
# Modifying the dataframe to perform analysis
count = 0
mylist = []
for row in extra_df.collect():
    temp = (count,list(filter(None, row)))
    mylist.append(temp)
    count = count + 1
    
# Creating a new dataframe from the above list
extra_df1 = spark.createDataFrame(mylist, ["id", "items"])

In [0]:
# Splitting the dataset into train and test dataframes
trainDF, testDF = extra_df1.randomSplit([0.8, 0.2], seed=25)
print(trainDF.cache().count()) # Cache because accessing training data multiple times
print(testDF.count())

7884
1951


In [0]:
# FP Growth Model
fpGrowth = FPGrowth(itemsCol="items", minSupport=0.01, minConfidence=0.01)
model = fpGrowth.fit(trainDF)

In [0]:
# Display frequent itemsets.
display(model.freqItemsets)

items freq List(margarine) 209 List(citrus fruit) 644 List(citrus fruit, root vegetables) 109 List(citrus fruit, sausage) 86 List(citrus fruit, other vegetables) 151 List(citrus fruit, tropical fruit) 154 List(citrus fruit, yogurt) 87 List(citrus fruit, whole milk) 148 List(white wine) 92 List(frozen meals) 153 List(bottled beer) 381 List(hard cheese) 114 List(newspapers) 244 List(whole milk) 1771 List(chocolate) 152 List(chicken) 334 List(chicken, other vegetables) 102 List(chicken, whole milk) 79 List(herbs) 110 List(coffee) 243 List(other vegetables) 1363 List(other vegetables, whole milk) 409 List(pip fruit) 591 List(pip fruit, citrus fruit) 103 List(pip fruit, root vegetables) 96 List(pip fruit, sausage) 85 List(pip fruit, other vegetables) 129 List(pip fruit, tropical fruit) 163 List(pip fruit, whole milk) 145 List(candy) 90 List(ham) 207 List(napkins) 106 List(shopping bags) 238 List(rolls/buns) 980 List(rolls/buns, other vegetables) 111 List(rolls/buns, whole milk) 181 List(oil) 89 List(UHT-milk) 206 List(bottled water) 480 List(bottled water, whole milk) 82 List(whipped/sour cream) 325 List(whipped/sour cream, whole milk) 90 List(butter milk) 151 List(packaged fruit/vegetables) 83 List(frozen vegetables) 194 List(frankfurter) 477 List(frankfurter, other vegetables) 84 List(frankfurter, whole milk) 104 List(ice cream) 148 List(butter) 310 List(butter, whole milk) 121 List(dessert) 238 List(tropical fruit) 831 List(tropical fruit, other vegetables) 191 List(tropical fruit, whole milk) 211 List(sugar) 105 List(specialty chocolate) 130 List(curd) 303 List(curd, whole milk) 117 List(soda) 813 List(soda, rolls/buns) 130 List(waffles) 103 List(berries) 237 List(white bread) 184 List(canned beer) 463 List(long life bakery product) 103 List(meat) 214 List(root vegetables) 795 List(root vegetables, other vegetables) 268 List(root vegetables, tropical fruit) 134 List(root vegetables, whole milk) 244 List(pork) 459 List(pork, root vegetables) 86 List(pork, other vegetables) 115 List(pork, whole milk) 107 List(grapes) 163 List(misc. beverages) 128 List(brown bread) 292 List(cream cheese) 163 List(pastry) 429 List(pastry, whole milk) 101 List(salty snack) 120 List(hamburger meat) 269 List(hamburger meat, other vegetables) 80 List(sliced cheese) 102 List(onions) 213 List(yogurt) 793 List(yogurt, rolls/buns) 94 List(yogurt, other vegetables) 144 List(yogurt, tropical fruit) 106 List(yogurt, whole milk) 218 List(domestic eggs) 244 List(specialty bar) 119 List(chewing gum) 97 List(fruit/vegetable juice) 211 List(sausage) 756 List(sausage, rolls/buns) 138 List(sausage, root vegetables) 93 List(sausage, other vegetables) 164 List(sausage, tropical fruit) 100 List(sausage, whole milk) 160 List(beverages) 158 List(beef) 411 List(beef, root vegetables) 108 List(beef, other vegetables) 99 List(beef, whole milk) 105

In [0]:
# Display generated association rules.
display(model.associationRules)

antecedent consequent confidence lift support List(other vegetables) List(citrus fruit) 0.1107850330154072 1.3562565221948297 0.01915271435819381 List(other vegetables) List(chicken) 0.0748349229640499 1.7664626726005073 0.01293759512937595 List(other vegetables) List(whole milk) 0.30007336757153336 1.3358432693020716 0.05187721968543886 List(other vegetables) List(pip fruit) 0.0946441672780631 1.2625628000342632 0.016362252663622526 List(other vegetables) List(rolls/buns) 0.08143800440205429 0.6551604354140775 0.014079147640791476 List(other vegetables) List(frankfurter) 0.061628760088041086 1.018618751643849 0.0106544901065449 List(other vegetables) List(tropical fruit) 0.1401320616287601 1.3294839637558902 0.024226281075596143 List(other vegetables) List(root vegetables) 0.1966250917094644 1.9499273245753679 0.033992897006595635 List(other vegetables) List(pork) 0.08437270726338958 1.4492253247593976 0.014586504312531709 List(other vegetables) List(hamburger meat) 0.058694057226705794 1.7202377218414442 0.010147133434804667 List(other vegetables) List(yogurt) 0.10564930300807043 1.0503645711420269 0.0182648401826484 List(other vegetables) List(sausage) 0.12032281731474688 1.254795094853789 0.020801623541349568 List(other vegetables) List(beef) 0.07263389581804842 1.393298381093659 0.012557077625570776 List(pastry) List(whole milk) 0.23543123543123542 1.0480744551890797 0.012810755961440892 List(whipped/sour cream) List(whole milk) 0.27692307692307694 1.2327846066976502 0.01141552511415525 List(yogurt) List(citrus fruit) 0.10970996216897856 1.3430952511494207 0.011035007610350075 List(yogurt) List(rolls/buns) 0.11853720050441362 0.953619682425303 0.011922881785895484 List(yogurt) List(other vegetables) 0.18158890290037832 1.0503645711420269 0.0182648401826484 List(yogurt) List(tropical fruit) 0.13366960907944514 1.2681723200750248 0.013444951801116184 List(yogurt) List(whole milk) 0.27490542244640603 1.2238025694903811 0.027650938609842718 List(rolls/buns) List(other vegetables) 0.11326530612244898 0.6551604354140775 0.014079147640791476 List(rolls/buns) List(whole milk) 0.1846938775510204 0.8222058332084952 0.02295788939624556 List(rolls/buns) List(soda) 0.1326530612244898 1.2863920475939454 0.016489091831557584 List(rolls/buns) List(yogurt) 0.09591836734693877 0.953619682425303 0.011922881785895484 List(rolls/buns) List(sausage) 0.14081632653061224 1.4685131195335277 0.01750380517503805 List(tropical fruit) List(citrus fruit) 0.18531889290012032 2.2687176267461937 0.019533231861998984 List(tropical fruit) List(pip fruit) 0.19614921780986763 2.6166504792098078 0.02067478437341451 List(tropical fruit) List(other vegetables) 0.22984356197352587 1.32948396375589 0.024226281075596143 List(tropical fruit) List(whole milk) 0.2539109506618532 1.1303410135618581 0.02676306443429731 List(tropical fruit) List(root vegetables) 0.16125150421179302 1.5991281247871398 0.016996448503297817 List(tropical fruit) List(yogurt) 0.12755716004813478 1.2681723200750248 0.013444951801116184 List(tropical fruit) List(sausage) 0.12033694344163658 1.2549424101770672 0.012683916793505834 List(chicken) List(other vegetables) 0.30538922155688625 1.766462672600507 0.01293759512937595 List(chicken) List(whole milk) 0.23652694610778444 1.052952254722627 0.010020294266869609 List(bottled water) List(whole milk) 0.17083333333333334 0.7605025409373236 0.010400811770674784 List(sausage) List(citrus fruit) 0.11375661375661375 1.3926353149955635 0.010908168442415017 List(sausage) List(pip fruit) 0.11243386243386243 1.4998791394730482 0.010781329274479959 List(sausage) List(rolls/buns) 0.18253968253968253 1.4685131195335277 0.01750380517503805 List(sausage) List(root vegetables) 0.12301587301587301 1.2199460916442049 0.011796042617960426 List(sausage) List(other vegetables) 0.21693121693121692 1.2547950948537887 0.020801623541349568 List(sausage) List(tropical fruit) 0.13227513227513227 1.2549424101770672 0.012683916793505834 List(sausage) List(whole milk) 0.2116

In [0]:
# transform examines the input items against all the association rules and summarize the
# consequents as prediction
display(model.transform(testDF))

id items prediction 13 List(frankfurter, rolls/buns, soda) List(other vegetables, whole milk, yogurt, sausage) 19 List(specialty bar) List() 25 List(bottled water, canned beer) List(whole milk) 37 List(canned beer) List() 44 List(butter milk, yogurt, cream cheese, spread cheese) List(citrus fruit, rolls/buns, other vegetables, tropical fruit, whole milk) 46 List(pastry, bottled water) List(whole milk) 56 List(packaged fruit/vegetables, brown bread, canned beer) List() 59 List(rolls/buns, pastry, sugar) List(whole milk, other vegetables, soda, yogurt, sausage) 60 List(other vegetables, whole milk, frozen vegetables, canned fish) List(citrus fruit, chicken, pip fruit, rolls/buns, frankfurter, tropical fruit, root vegetables, pork, hamburger meat, yogurt, sausage, beef, bottled water, whipped/sour cream, butter, curd, pastry) 65 List(whole milk) List(citrus fruit, chicken, other vegetables, pip fruit, rolls/buns, bottled water, whipped/sour cream, frankfurter, butter, tropical fruit, curd, root vegetables, pork, pastry, yogurt, sausage, beef) 66 List(curd cheese, coffee) List() 70 List(red/blush wine) List() 96 List(sausage, sliced cheese, rolls/buns, brown bread) List(other vegetables, whole milk, soda, yogurt, citrus fruit, pip fruit, root vegetables, tropical fruit) 103 List(sausage, whole milk, yogurt, coffee) List(citrus fruit, rolls/buns, other vegetables, tropical fruit, pip fruit, root vegetables, chicken, bottled water, whipped/sour cream, frankfurter, butter, curd, pork, pastry, beef) 106 List(beef, citrus fruit, berries, root vegetables) List(sausage, other vegetables, tropical fruit, yogurt, whole milk, pip fruit, pork) 110 List(frankfurter, sausage, root vegetables, other vegetables) List(citrus fruit, chicken, whole milk, pip fruit, rolls/buns, tropical fruit, pork, hamburger meat, yogurt, beef) 112 List(whole milk, dessert, frozen vegetables, rolls/buns) List(other vegetables, soda, yogurt, sausage, citrus fruit, chicken, pip fruit, bottled water, whipped/sour cream, frankfurter, butter, tropical fruit, curd, root vegetables, pork, pastry, beef) 120 List(frankfurter, sausage, long life bakery product, waffles) List(citrus fruit, pip fruit, rolls/buns, root vegetables, other vegetables, tropical fruit, whole milk) 123 List(other vegetables, red/blush wine) List(citrus fruit, chicken, whole milk, pip fruit, rolls/buns, frankfurter, tropical fruit, root vegetables, pork, hamburger meat, yogurt, sausage, beef) 134 List(curd, flour, liquor (appetizer)) List(whole milk) 144 List(citrus fruit, butter, cream cheese, white bread) List(root vegetables, sausage, other vegetables, tropical fruit, yogurt, whole milk, pip fruit) 145 List(frankfurter, rolls/buns) List(other vegetables, whole milk, soda, yogurt, sausage) 150 List(other vegetables, whole milk, meat spreads, cake bar) List(citrus fruit, chicken, pip fruit, rolls/buns, frankfurter, tropical fruit, root vegetables, pork, hamburger meat, yogurt, sausage, beef, bottled water, whipped/sour cream, butter, curd, pastry) 156 List(popcorn) List() 160 List(whole milk, butter, dessert, rolls/buns) List(other vegetables, soda, yogurt, sausage, citrus fruit, chicken, pip fruit, bottled water, whipped/sour cream, frankfurter, tropical fruit, curd, root vegetables, pork, pastry, beef) 174 List(curd, dessert, soda, salty snack) List(whole milk, rolls/buns) 180 List(ham, whole milk, rolls/buns) List(other vegetables, soda, yogurt, sausage, citrus fruit, chicken, pip fruit, bottled water, whipped/sour cream, frankfurter, butter, tropical fruit, curd, root vegetables, pork, pastry, beef) 181 List(other vegetables, ice cream) List(citrus fruit, chicken, whole milk, pip fruit, rolls/buns, frankfurter, tropical fruit, root vegetables, pork, hamburger meat, yogurt, sausage, beef) 184 List(herbs, candy) List() 191 List(specialty bar) List() 192 List(curd, salt, liquor (appetizer), detergent) List(whole milk) 193 List(frankfurter, rolls/buns) List(other vegetables, whole milk, soda, yogurt,

<h3> 1.9 References: </h3>

<p>
  <b>1.</b>  Dr. Liao’s Code Examples & Tutorials: Blackboard/Liao_PySpark_basic_databricks.html
  <br>
  <b>2.</b> PySpark: https://spark.apache.org/docs/2.4.0/api/python/pyspark.html  
  <br>
  <b>3.</b> Frequent Pattern Mining : https://spark.apache.org/docs/latest/ml-frequent-pattern-mining.html
</p>